distsql/select_result.go

```diff
@@ -85,74 +77,51 @@ type selectResult struct {
 }
 
 func (r *selectResult) Fetch(ctx context.Context) {
-	go r.fetch(ctx)
 }
 
-func (r *selectResult) fetch(ctx context.Context) {
-	startTime := time.Now()
-	defer func() {
-		if c := recover(); c != nil {
-			err := fmt.Errorf("%v", c)
-			logutil.Logger(ctx).Error("OOM", zap.Error(err))
-			r.results <- resultWithErr{err: err}
-		}
-
-		close(r.results)
-		duration := time.Since(startTime)
-		// TODO: Add a label to distinguish between success or failure.
-		// https://github.com/pingcap/tidb/issues/11397
-		metrics.DistSQLQueryHistgram.WithLabelValues(r.label, r.sqlType).Observe(duration.Seconds())
-	}()
-	for {
-		var result resultWithErr
-		resultSubset, err := r.resp.Next(ctx)
-		if err != nil {
-			result.err = err
-		} else if resultSubset == nil {
-			// If the result is drained, the resultSubset would be nil
-			return
-		} else {
-			result.result = resultSubset
-			r.memConsume(int64(resultSubset.MemSize()))
-		}
-
-		select {
-		case r.results <- result:
-		case <-r.closed:
-			// If selectResult called Close() already, make fetch goroutine exit.
-			if resultSubset != nil {
-				r.memConsume(-int64(resultSubset.MemSize()))
-			}
-			return
-		case <-ctx.Done():
-			if resultSubset != nil {
-				r.memConsume(-int64(resultSubset.MemSize()))
-			}
-			return
-		}
+func (r *selectResult) fetchResp(ctx context.Context) error {
+	r.respChkIdx = 0
+	resultSubset, err := r.resp.Next(ctx)
+	if err != nil {
+		return errors.Trace(err)
 	}
-}
-
-// NextRaw returns the next raw partial result.
-func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
-	re := <-r.results
+	if resultSubset == nil {
+		r.selectResp = nil
+		return nil
+	}
+	r.selectResp = new(tipb.SelectResponse)
+	err = r.selectResp.Unmarshal(resultSubset.GetData())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	r.selectRespSize = r.selectResp.Size()
```

Comment:
We should keep tracking the memory of `r.selectResp()`.

Reply:
Updated, PTAL